### DISCUSSION TITLE
minimum separating space between completion and its annotation?

### DISCUSSION THREAD

**Discussion by ville-h** [company-mode/company-mode#1376] [Q&A] [answered]
Is there a setting to force at least N spaces between the completion and its annotation? Poor ascii art follows:

```
+--------------------------------------------------+
| completion0completion0 <forced space> annotation0|
| completion1 <padding..><forced space> annotation1|
+--------------------------------------------------+
```
So the `<forced space>` would be what I am looking for.

**Comment by yugaego** [comment, marked as answer]
Hi and welcome. Can we also assume enabling [company-tooltip-align-annotations](http://company-mode.github.io/manual/Frontends.html#index-company_002dtooltip_002dalign_002dannotations) with

```lisp
(setq company-tooltip-align-annotations t)
```

isn't a sufficient solution for you?

**Reply by ville-h** [reply]
Yeah I've `company-tooltip-align-annotations` set. I wish to have the forced minimum separation in addition to that. Also setting `company-tooltip-minimum-width` isn't enough because of varying length of the completions at any given time.

**Reply by dgutov** [reply]
Hi!

Here's a patch you can try:

```diff
diff --git a/company.el b/company.el
index b95c4d8..7c4d654 100644
--- a/company.el
+++ b/company.el
@@ -300,8 +300,12 @@ This doesn't include the margins and the scroll bar."
                  (const :tag "Two lines" lines)))
 
 (defcustom company-tooltip-align-annotations nil
-  "When non-nil, align annotations to the right tooltip border."
-  :type 'boolean
+  "When non-nil, align annotations to the right tooltip border.
+
+When the value is a number, maintain at least this many spaces between the
+completion text and its annotation."
+  :type '(choice (const :tag "Align to the right" t)
+                 (number :tag "Align to the right with minimum spacing"))
   :package-version '(company . "0.7.1"))
 
 (defcustom company-tooltip-flip-when-above nil
@@ -3087,11 +3091,11 @@ If SHOW-VERSION is non-nil, show the version in the echo area."
          (ann-ralign company-tooltip-align-annotations)
          (ann-truncate (< width
                           (+ (length value) (length annotation)
-                             (if ann-ralign 1 0))))
+                             (or ann-ralign 0))))
          (ann-start (+ margin
                        (if ann-ralign
                            (if ann-truncate
-                               (1+ (length value))
+                               (+ (length value) ann-ralign)
                              (- width (length annotation)))
                          (length value))))
          (ann-end (min (+ ann-start (length annotation)) (+ margin width)))
@@ -3099,7 +3103,8 @@ If SHOW-VERSION is non-nil, show the version in the echo area."
                        (if (or ann-truncate (not ann-ralign))
                            (company-safe-substring
                             (concat value
-                                    (when (and annotation ann-ralign) " ")
+                                    (when (and annotation ann-ralign)
+                                      (company-space-string ann-ralign))
                                     annotation)
                             0 width)
                          (concat
@@ -3327,6 +3332,10 @@ but adjust the expected values appropriately."
 (defun company--create-lines (selection limit)
   (let ((len company-candidates-length)
         (window-width (company--window-width))
+        (company-tooltip-align-annotations
+         (if (eq company-tooltip-align-annotations t)
+             1
+           company-tooltip-align-annotations))
         left-margins
         left-margin-size
         lines
@@ -3400,7 +3409,8 @@ but adjust the expected values appropriately."
         (push (list value annotation left) items)
         (setq width (max (+ (length value)
                             (if (and annotation company-tooltip-align-annotations)
-                                (1+ (length annotation))
+                                (+ (length annotation)
+                                   company-tooltip-align-annotations)
                               (length annotation)))
                          width))))
 
```

**Reply by ville-h** [reply]
Thank you, that's great! Unfortunately I am not very familiar with how emacs addons work. I am only qualified to copy paste a `use-package`-snippet and that's about it. What is the process for testing a patch like this? Just go in `~/.config/emacs/packages/company...`, apply the patch on top of `company.el` and go?

**Reply by ville-h** [reply]
Looks like I managed it. Created a buffer with the contents of your patch, execute `diff-ediff-patch`, pointed it at `~/.config/emacs/.../company.el` and quit the confusing interface. Adjusted my `company-tooltip-align-annotations` to 4 and restarted. It does appear to do what I was asking for. Thank you.

**Reply by dgutov** [reply]
Excellent.

> I am only qualified to copy paste a use-package-snippet and that's about it.

And now you are qualified for much more! ;-)

> Looks like I managed it. Created a buffer with the contents of your patch, execute diff-ediff-patch, pointed it at ~/.config/emacs/.../company.el and quit the confusing interface.

Two other options to do the same:

1. Paste the patch into a new buffer and switch it to `diff-mode`. Then iteratively apply it using `C-c C-a` (diff-mode might ask you at first where the file to patch is), until you reach the end of the patch buffer. Then switch to the patched file and save. You could also re-evaluate changed definitions using `C-M-x` (that command evaluates the one around point) instead of restarting.
2. Or just save the contents to a new `foo.diff` and then apply it using the `patch` program from the command-line. This one is sometimes less forgiving towards whitespacing problems than the previous option.

**Reply by dgutov** [reply]
Anyway, thanks for testing, I've pushed that change to master now.

I had a brief doubt whether we'd rather have a separate setting, so that left-aligned annotations could have padding too, but that could probably be retrofitted later somehow, if need arises.

**Reply by yugaego** [reply]
> I had a brief doubt whether we'd rather have a separate setting, so that left-aligned annotations could have padding too

I had that on mind too: a separate setting for annotations left padding, no matter what alignment is. Left-aligned annotations with additional spacing on the left could be visually/UX preferable option, I think.

**Reply by dgutov** [reply]
Do you think it will be useful? I checked out a bunch of editors yesterday (screenshots and vids), and none seemed to add spacing before method arguments -- if they showed those at all.

It wouldn't be too difficult to support both, though.

**Reply by yugaego** [reply]
> Do you think it will be useful?

For someone, probably. Supposedly, I'd use left alignment  of annotations with a bit of spacing myself. This thread made me realize I use right alignment mostly (maybe only) because it adds spacing between candidate and annotation. For me, additional spacing makes it easier/faster to look through the candidates.
   
> It wouldn't be too difficult to support both

Nice to know it. I started looking into it, being an interested part, but had to pause/stop for now.

**Reply by dgutov** [reply]
All right, I've split the var into a separate one, so it affects the left-aligned ones too. Check out the latest master.

Sorry to spoil the experimentation opportunity ;-)

**Reply by yugaego** [reply]
This is looking great, thanks!

> Sorry to spoil the experimentation opportunity

Oh, no problem, they're endless you know ;-)
[Seems it's 'bout time to adjust the manual.]
